socket.go

```diff
@@ -87,21 +89,24 @@ func newSocket(ctx context.Context, sockType SocketType, opts ...Option) *socket
 
 // Close closes the open Socket
 func (sck *socket) Close() error {
+	sck.closeMu.Lock()
+	defer sck.closeMu.Unlock()
+	defer close(sck.closedConns)
+
+	// We cancel() while closeMu is acquired to synchronize with scheduleRmConn
 	sck.cancel()
-	defer func() {
-		close(sck.closedConns)
-	}()
 
 	if sck.listener != nil {
 		defer sck.listener.Close()
 	}
 
+	sck.mu.RLock()
 	if sck.conns == nil {
+		sck.mu.RUnlock()
```

Comment:
perhaps turn that into a `defer` and move it after the `RLock` (and remove the `RUnlock` at line 116).

with Go-1.14, the defers will be really cheap.